File: sol/solution/0201-0300/0202/Solution.php
class Solution {
    /**
     * @param Integer $n
     * @return Boolean
     */
    function isHappy($n) {
        $getNext = function($num) {
            $sum = 0;
            while ($num > 0) {
                $d = $num % 10;
                $sum += $d * $d;
                $num = intdiv($num, 10);
            }
            return $sum;
        };
        $slow = $n;
        $fast = $getNext($n);
        while ($fast != 1 && $slow != $fast) {
            $slow = $getNext($slow);
            $fast = $getNext($getNext($fast));
        }
        return $fast == 1;
    }
}

File: sol/solution/0201-0300/0202/Solution.rb
# @param {Integer} n
# @return {Boolean}
def is_happy(n)
  get_next = ->(num) {
    num.digits.map { |d| d * d }.sum
  }
  slow = n
  fast = get_next.call(n)
  while fast != 1 && slow != fast
    slow = get_next.call(slow)
    fast = get_next.call(get_next.call(fast))
  end
  fast == 1
end
